### PR TITLE
Extract (beta)

### PR DESCRIPTION
**Pull request by nickscamara**

### CODE REVIEW

File: apps/api/src/controllers/v1/extract.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import { generateOpenAICompletions } from "../../scraper/scrapeURL/transformers/llmExtract";
+import { isUrlBlocked } from "../../scraper/WebScraper/utils/blocklist";
+import { getMapResults } from "./map";
+import { buildDocument } from "../../lib/extract/build-document";
+
+configDotenv();
+const redis = new Redis(process.env.REDIS_URL!);
+
+const MAX_EXTRACT_LIMIT = 100;
+const MAX_RANKING_LIMIT = 10;
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts:32]: Gotta figure out how to optimize the map ranking limit

Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts:32]: Fine for now

File: apps/api/src/controllers/v1/extract.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+const SCORE_THRESHOLD = 0.75;
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts]: same with the score

File: apps/api/src/controllers/v1/types.ts [hunk 3]
```diff
@@ -151,8 +151,25 @@ export const scrapeOptions = z.object({
 }).strict(strictMessage)
 
 
+
 export type ScrapeOptions = z.infer<typeof scrapeOptions>;
 
+export const extractV1Options = z.object({
+  urls: url.array(),
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/types.ts:158]: - [x] we need the url validation

File: apps/api/src/controllers/v1/extract.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+      
+      // Filter by path prefix if present
+      // wrong
+      // if (pathPrefix) {
+      //   mappedLinks = mappedLinks.filter(x => x.url && x.url.includes(`/${pathPrefix}/`));
+      // }
+
+      if (req.body.prompt) {
+        const linksAndScores : { link: string, linkWithContext: string, score: number, originalIndex: number }[] = await performRanking(mappedLinksRerank, mappedLinks.map(l => l.url), mapUrl);
+        mappedLinks = linksAndScores
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts]: we might want unit testing for the ranking

Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts]: - [x] add comment explaining what these filters do

File: apps/api/src/controllers/v1/extract.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+  // Wait for all URL processing to complete and flatten results
+  const processedUrls = await Promise.all(urlPromises);
+  links.push(...processedUrls.flat());
+
+  // console.log("links", links.length);
+  // Scrape all links in parallel
+  const scrapePromises = links.map(async (url) => {
+    const origin = req.body.origin || "api";
+    const timeout = req.body.timeout ?? 30000;
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts]: Maybe rename this timeout to scrapeTimeout?

File: apps/api/src/controllers/v1/extract.ts [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+      team_id: req.auth.team_id,
+      basePriority: 10,
+    });
+
+    await addScrapeJob(
+      {
+        url,
+        mode: "single_urls", 
+        team_id: req.auth.team_id,
+        scrapeOptions: scrapeOptions.parse({}),
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts:162]: - [x] if we rename the timeout need to fix the parsing here to include it

File: apps/api/src/controllers/v1/extract.ts [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        plan: req.auth.plan!,
+        origin,
+        is_scrape: true,
+      },
+      {},
+      jobId,
+      jobPriority
+    );
+
+    const totalWait = 0;
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts]: doens't do anything

File: apps/api/src/controllers/v1/extract.ts [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+  // console.log("completions", completions);
+
+  // if(req.body.extract && req.body.formats.includes("extract")) {
+  //   creditsToBeBilled = 5;
+  // }
+
+  // TODO: change this later
+  // While on beta, we're billing 5 credits per link discovered/scraped.
+  billTeam(req.auth.team_id, req.acuc?.sub_id, links.length * 5).catch(error => {
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts:219]: - [ ] Need to improve this

File: apps/api/src/controllers/v1/extract.ts [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+  });
+
+
+  // console.log("completions.extract", completions.extract);
+
+  let data: any;
+  try {
+    data = JSON.parse(completions.extract);
+  } catch (e) {
+    data = completions.extract;
```
Comment (nickscamara) [inline apps/api/src/controllers/v1/extract.ts]: - [x] hmm, might wanna throw an error? let the user know?

File: apps/api/src/routes/v1.ts [hunk 10]
```diff
@@ -178,6 +179,14 @@ v1Router.ws(
     crawlStatusWSController
 );
 
+v1Router.post(
+    "/extract",
+    authMiddleware(RateLimiterMode.Scrape),
+    checkCreditsMiddleware(1),
+    blocklistMiddleware,
```
Comment (nickscamara) [inline apps/api/src/routes/v1.ts]: - [x] will the blocklist middle ware even work here? if not remove it.

File: apps/api/src/scraper/scrapeURL/engines/index.ts [hunk 11]
```diff
@@ -6,14 +6,17 @@ import { scrapePDF } from "./pdf";
...
-export type Engine = "fire-engine;chrome-cdp" | "fire-engine;playwright" | "fire-engine;tlsclient" | "scrapingbee" | "scrapingbeeLoad" | "playwright" | "fetch" | "pdf" | "docx";
+export type Engine = "fire-engine;chrome-cdp" | "fire-engine;playwright" | "fire-engine;tlsclient" | "scrapingbee" | "scrapingbeeLoad" | "playwright" | "fetch" | "pdf" | "docx" | "cache";
 
 const useScrapingBee = process.env.SCRAPING_BEE_API_KEY !== '' && process.env.SCRAPING_BEE_API_KEY !== undefined;
 const useFireEngine = process.env.FIRE_ENGINE_BETA_URL !== '' && process.env.FIRE_ENGINE_BETA_URL !== undefined;
 const usePlaywright = process.env.PLAYWRIGHT_MICROSERVICE_URL !== '' && process.env.PLAYWRIGHT_MICROSERVICE_URL !== undefined;
+const useCache = process.env.CACHE_REDIS_URL !== '' && process.env.CACHE_REDIS_URL !== undefined;
 
 export const engines: Engine[] = [
+    ...(useCache ? [ "cache" as const ] : []),
```
Comment (nickscamara) [inline apps/api/src/scraper/scrapeURL/engines/index.ts]: - [x] let's not have this for now

File: apps/api/src/services/queue-jobs.ts
Comment (nickscamara) [inline apps/api/src/services/queue-jobs.ts:1]: - [ ] can we affor this rate? let's hope so! @mogery

File: apps/js-sdk/firecrawl/src/index.ts
Comment (nickscamara) [inline apps/js-sdk/firecrawl/src/index.ts:1]: need testing

File: apps/python-sdk/firecrawl/firecrawl.py
Comment (nickscamara) [inline apps/python-sdk/firecrawl/firecrawl.py:1]: need testing

File: apps/api/src/controllers/v1/types.ts (hunk 3 above)
Comment (nickscamara) [reply apps/api/src/controllers/v1/types.ts:158]: Done!

File: apps/api/src/controllers/v1/extract.ts (hunk 4 above)
Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts]: Done!

File: apps/api/src/controllers/v1/extract.ts (hunk 5 above)
Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts]: Lets not expose this timeout and instead just make a % of the total request timeout.

Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts]: Done!

File: apps/api/src/controllers/v1/extract.ts (hunk 7 above)
Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts]: Fixed!!

File: apps/api/src/services/queue-jobs.ts
Comment (nickscamara) [reply apps/api/src/services/queue-jobs.ts:1]: Pushing for 250 for now as I commented out cache layer

File: apps/api/src/controllers/v1/extract.ts [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+      if (req.body.prompt) {
+        // Get similarity scores between the search query and each link's context
+        const linksAndScores = await performRanking(mappedLinksRerank, mappedLinks.map(l => l.url), mapUrl);
+        
+        // First try with high threshold
+        let filteredLinks = filterAndProcessLinks(mappedLinks, linksAndScores, INITIAL_SCORE_THRESHOLD);
+        
+        // If we don't have enough high-quality links, try with lower threshold
+        if (filteredLinks.length < MIN_REQUIRED_LINKS) {
+          console.log(`Only found ${filteredLinks.length} links with score > ${INITIAL_SCORE_THRESHOLD}. Trying lower threshold...`);
```
Comment (mogery) [inline apps/api/src/controllers/v1/extract.ts]: only `logger` pls

File: apps/api/src/lib/extract/completions.ts [hunk 13]
```diff
@@ -0,0 +1,124 @@
+// use llmExtract.ts instead
```
Comment (mogery) [inline apps/api/src/lib/extract/completions.ts:1]: can we just delete this then?

**Review by mogery — APPROVED** [review]
LGTM

File: apps/api/src/lib/extract/completions.ts (hunk 13 above)
Comment (nickscamara) [reply apps/api/src/lib/extract/completions.ts:1]: yep

File: apps/api/src/controllers/v1/extract.ts (hunk 12 above)
Comment (nickscamara) [reply apps/api/src/controllers/v1/extract.ts]: done

**Pull request merged**